### PR TITLE
refactor: rasa/shared/core/training_data/visualization.py

### PR DESCRIPTION
**Pull request by arjun-234**
In the refactored version, In function(_remove_auxiliary_nodes), I've replaced the conversion of graph.predecessors(i) into a list with the direct usage of the generator. Additionally, I've introduced the predecessors_seen set to efficiently keep track of seen predecessors. When a duplicated predecessor is found, we can remove the node and break out of the inner loop.
**Proposed changes**:
- This optimization reduces the time complexity of checking for duplicated nodes to approximately O((TMP_NODE_ID - special_node_idx) + out_degree(node)).

**Status (please check what you already did)**:
- [x] Implemented performance optimizations

### CODE REVIEW

**Review by Urkem — COMMENTED** [review]
💯 Thanks for the contribution can you also add a unit test for `_remove_auxiliary_nodes` [here](https://github.com/RasaHQ/rasa/blob/main/tests/shared/core/training_data/test_visualization.py)

**Comment by arjun-234** [comment]
> 💯 Thanks for the contribution can you also add a unit test for `_remove_auxiliary_nodes` [here](https://github.com/RasaHQ/rasa/blob/main/tests/shared/core/training_data/test_visualization.py)

Hey @Urkem,
Thank you for acknowledging my contribution! I have implemented unit test for the function: _remove_auxiliary_nodes. Feel free to review it whenever you have a moment.
Thank you once again!

File: tests/shared/core/training_data/test_visualization.py [hunk 1]
```diff
@@ -188,3 +188,21 @@ def test_story_visualization_with_merging(domain: Domain):
     assert 15 < len(generated_graph.nodes()) < 33
 
     assert 20 < len(generated_graph.edges()) < 33
+
+
+def test_remove_auxiliary_nodes():
```
Comment (Urkem) [inline tests/shared/core/training_data/test_visualization.py:193]: Can you add 2-3 example test graphs and put them in `@pytest.mark.parametrize`

File: tests/shared/core/training_data/test_visualization.py [hunk 2]
```diff
@@ -188,3 +188,21 @@ def test_story_visualization_with_merging(domain: Domain):
...
+    # Create a sample graph
+    graph = nx.MultiDiGraph()
```
Comment (Urkem) [inline tests/shared/core/training_data/test_visualization.py:195]: `nx` is not defined just add `from networkx import MultiDiGraph`

File: tests/shared/core/training_data/test_visualization.py [hunk 3]
```diff
@@ -188,3 +188,21 @@ def test_story_visualization_with_merging(domain: Domain):
...
+    graph.add_nodes_from([-2, -1, 0, 1, 2, 3, 4, 5])
+    graph.add_edges_from([(-2, 0), (-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)])
+
+    # Call the method to remove auxiliary nodes
+    visualization._remove_auxiliary_nodes(graph, 3)
```
Comment (Urkem) [inline tests/shared/core/training_data/test_visualization.py:200]: The test fails with `networkx.exception.NetworkXError: The node -1 is not in the digraph.` 🤔  didn't investigate much please explain why it was happening.

**Review by Urkem — COMMENTED** [review]
Thanks for adding the unit test 👍

File: tests/shared/core/training_data/test_visualization.py (hunk 1 above)
Comment (arjun-234) [reply tests/shared/core/training_data/test_visualization.py]: @Urkem Done!

File: tests/shared/core/training_data/test_visualization.py (hunk 2 above)
Comment (arjun-234) [reply tests/shared/core/training_data/test_visualization.py:215]: @Urkem, Done!

File: tests/shared/core/training_data/test_visualization.py (hunk 3 above)
Comment (arjun-234) [reply tests/shared/core/training_data/test_visualization.py]: Hello @Urkem,
I have implemented two example test cases for the same. Could you please review them whenever you have a moment? Let me know if you are still facing any issues while testing. 
Thank you!

Comment (Urkem) [reply tests/shared/core/training_data/test_visualization.py]: @arjun-234 Thanks for the new test case, they both pass now 👍

**Review by Urkem — APPROVED** [review]
💯 Thanks for the contribution just fix the the [ruff errors](https://github.com/RasaHQ/rasa/actions/runs/5507512493/jobs/10037772054?pr=12544)

**Comment by arjun-234** [comment]
> 💯 Thanks for the contribution just fix the the [ruff errors](https://github.com/RasaHQ/rasa/actions/runs/5507512493/jobs/10037772054?pr=12544)

Hey @Urkem , Done!

File: tests/shared/core/training_data/test_visualization.py [hunk 4]
```diff
@@ -188,3 +190,42 @@ def test_story_visualization_with_merging(domain: Domain):
...
+            set([0, 1, 2, 3, 4, 5, -1]),
+            [(-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
+        ),
+        (
+            [-3, -2, -1, 0, 1, 2, 3, 4, 5],
+            [(-3, -2), (-2, -1), (-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)],
+            4,
+            set([-3, -1, 0, 1, 2, 3, 4, 5]),
+            [(-1, 0), (0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
+        )
```
Comment (Urkem) [inline tests/shared/core/training_data/test_visualization.py:211]: You only have test cases when the graph is connected with only one edge between subsequent nodes. While the original function will remove all the predecessors nodes if it was already in `predecessors_seen` the refactored version will just remove the first one and continue. Please keep the same functionality after the refactor 🙏

**Review by Urkem — CHANGES_REQUESTED** [review]
Just one small comment :)

File: tests/shared/core/training_data/test_visualization.py (hunk 4 above)
Comment (arjun-234) [reply tests/shared/core/training_data/test_visualization.py:211]: Hey @Urkem,
I have removed the break statement from the loop to ensure that all predecessor nodes present in the "predecessors_seen" collection are removed without exiting the loop. Now, This refactor maintains the same functionality as before. Please review it :)
Thank You!

**Review by Urkem — APPROVED** [review]

**Comment by arjun-234** [comment]
Hey @Urkem, 
I've addressed all the code style issues and made necessary fixes. Could you please review it once you are available.
Thanks :)

**Comment by arjun-234** [comment]
Hey @Urkem,
Please review it once you are available :)

**Comment by vcidst** [comment]
Hey @arjun-234 I think these changes are making a regression test fail at, `tests/regressions/test_action_two_stage_fallback_11294.py::test_action_two_stage_fallback_does_not_return_key_error` do you think something in the test can be changed to avoid this failure? I've enabled auto-merge for this PR so it will be merged once the tests pass

**Comment by arjun-234** [comment]
Hey @vcidst,
I'm puzzled by the failures. The first occurrence failed during [Run Tests (test-other-unit-tests, ubuntu-22.04, 3.8)](https://github.com/RasaHQ/rasa/actions/runs/5784910005/job/15701264549#logs). Unfortunately, we're facing different failure now in [Run Tests (test-other-unit-tests, windows-2019, 3.10)](https://github.com/RasaHQ/rasa/actions/runs/5794802858/job/15706540836?pr=12544#logs) :(

**Comment by arjun-234** [comment]
> @arjun-234 I see, it might be a flaky test in that case. I'll re-run the CI and see if the problem persists

@vcidst Thanks :)

**Pull request merged**